content/posts/2023/kerberos-for-the-whole-workflow.md

```diff
@@ -0,0 +1,130 @@
+        commands:
+          - ...
+      - name: mysecondstep
+        environment: ...
+        commands:
+          - ...
+      - name: publish
+        kerberos: true
+        environment: "reanahub/reana-auth-krb5:1.0.1"
+        commands:
+          - mkdir -p /eos/home-j/johndoe/myoutputs/myplots
+          - cp myplots/*.png /eos/home-j/johndoe/myoutputs/myplots
+```
+
+Note that you would need to specify the `kerberos: true` workflow hint for each
+step of the workflow where the Kerberos authentication was necessary. This
+approach works well, but it may be tedious if you need to add the Kerberos hint
+for many steps in your workflow.
+
+## Enable Kerberos at the workflow orchestration level
+
+As of REANA 0.9.0, it is now possible to enable Kerberos at the workflow
+orchestration level itself by setting the `kerberos: true` hint in the
+`reana.yaml`
+[`resources`](https://docs.reana.io/reference/reana-yaml/#reanayaml-workflow)
+clause.
+
+As a realistic example of where this feature may come useful, let us consider
+an analysis using an unknown number of input files located on a remote server
+that is only accessible with Kerberos credentials, for example on
+`eosuser.cern.ch`:
+
+```console 
+$ xrdfs root://eosuser.cern.ch ls -l /eos/user/j/johndoe/mydata/
+---- 2023-01-23 15:10:24       12345 /eos/user/j/johndoe/mydata/myfile_1.csv
+---- 2023-01-23 15:10:31       23456 /eos/user/j/johndoe/mydata/myfile_2.csv
+---- 2023-01-23 15:10:35       34567 /eos/user/j/johndoe/mydata/myfile_3.csv
+...
+```
+
+Let us write a Snakemake workflow that will fetch these input data files
+concurrently irrespective of their number as the first step of our workflow.
+First, we shall include the above mentioned `resources` clause in the
+`reana.yaml` workflow specification file:
+
+```yaml {hl_lines=[9,10]}
+inputs:
+  files:
+    - Snakefile
+  parameters:
+    input: inputs.yaml
+workflow:
+  type: snakemake
+  file: Snakefile
+  resources:
+    kerberos: true
+outputs:
+  files:
+  - myoutput.png
+```
+
+The `Snakefile` workflow definition can now rely on Kerberos being available,
+and we can define the first `fetch_data` step using XRootD remote provider as
+follows:
+
+```python
+from snakemake.remote.XRootD import RemoteProvider as XRootDRemoteProvider
+
+XRootD = XRootDRemoteProvider(stay_on_remote=True)
+
+file_numbers = XRootD.glob_wildcards("root://eosuser.cern.ch//eos/user/j/johndoe/mydata/myfile_{n}.csv").n
+
+rule all:
+    input:
+        expand("mylocaldata/myfile_{n}.csv", n=file_numbers)
+
+rule fetch_data:
+    input:
+        XRootD.remote("root://eosuser.cern.ch//eos/user/j/johndoe/mydata/myfile_{n}.csv")
+    output:
+        "mylocaldata/myfile_{n}.csv"
+    container:
+        "docker://opensciencegrid/osgvo-el7:release-20211029-0011"
+    shell:
+        "xrdcp {input[0]} {output[0]} && mkdir -p mylocaldata && touch {output[0]}"
+```
+
+(Note that currently you need to have a valid Kerberos ticket at the time of
```

Comment:
Amended everything with the example and verbiage.

This part may be a bit tricky for readers, who may wonder why do we require Kerberos ticket locally also at the time of _submission_ of the workflow on the client side, and not only at the time of _start_ of the workflow on the server side. Would feel more natural that way.

However, in order to change this behaviour, we would have to amend the client-side `create` command to only submit workflow definition, and let the server-side handle everything. But we are not there yet in the thin-client sprint.

Something to ponder whether to postpone or not this example for that sprint...